### ISSUE TITLE
Consider declaring methods rather than using magic method __call() in src/Factory

### ISSUE THREAD

**Issue by apotek** [ChromaticHQ/orange-dam-php#23]
## Description

General statement: Static analysers (as well as code checkers) tend to become disoriented when magic methods are used. In addition, [many people consider them bad practice](https://www.php.net/manual/en/language.oop5.overloading.php#object.call), unless there is a genuine need for a contract-less, flexible object that can flexibly handle unanticipated needs from the end-user. 

The advantage, for example, of using `__call`  on Factory in this specific case, is that we can flexibly allow the end user to use get any EndPoint they might need, without the code having too _follow_ the API development, and create a (long long) list of all the endpoints. 

However, if we are comfortable limiting the library to work with a limited set of endpoints, it would be better practice to declare those methods in the class. 

## Options
1. Leave as-is. We might decide we don't want to support every single endpoint made available by Orange Dam. 
2. Implement a limited set of end-points that are useful for most common scenarios (but do we know which ones they are)?
3. Add whatever type-hinting/declarations that could at least help syntax checkers and static analyzers to handle Factory correctly. 

## Tertiary Considerations
If we refactor this library, I would also encourage it to be renamed from Factory to EndpointFactory.

**Comment by apotek** [comment]
Potential list of endpoints we would need to document/handle if we moved away from magic methods and still wanted to claim to support Orange Logic API.

application
asynchronousprocessing
asyncprocessing
automations
cms
configuration
crm
document
extensibility
fieldsecurity
finance
framework
indexing
infrastructure
integration
invisiblehand
mediafile
messaging
monitoring
objectmanagement
parameter
project
recyclebinupgrade
report
sample
search
security
share
taskmanager
tools
ui
upload
workflows
version1

**Comment by apotek** [comment]
Adding `@method` PHP docs to orange-dam-php library Factory does not make PHPStan smile. 

Looks like we will either have to implement the endpoints (and then keep up with them), or tell users of this library to use their own stub file in order to pass PHPStan's requirements. Of those two options, I favor adding a section to the README for telling users how to implement a custom stub file https://phpstan.org/user-guide/stub-files. Another means would be to encourage end users to use a Mixin rather than the Factory directly. According to PHPStan docs, the @mixin tag should help, but I haven't built a proof case for that.

**Comment by apotek** [comment]
After further discussion with @markdorison and getting more familiar with the architecture of the code, we have decided to remove the magic `__call()` and implement methods for each endpoint. Given that each endpoint class (in the Endpoints) directory are not able to be abstracted into truly dynamic "Endpoint" classes, we, _de _facto_ already only support those endpoints for which we have created a class in the Endpoints directory. So we do not in fact gain any of the flexibility that the use of the `__call()` magic should allow us. 

A further gain by explicitly creating method calls for each endpoint is that any user of the class can use PHPStan without having to create ignore rules for this library.
